lib/corefoundation/preferences.rb

```diff
@@ -13,55 +10,120 @@ module CF
+      #
+      # @return [VALUE] Preference value returned from the {#self.get} method call.
+      #
       def get!(key, application_id, username = nil, hostname = nil)
-        get(key, application_id, username, hostname) || (raise CF::Error, "Preference key `#{key}` not found")
+        value = get(key, application_id, username, hostname)
+        if value.nil?
+          raise(CF::Exceptions::PreferenceDoesNotExist.new(key, application_id, hostname))
+        else
+          value
+        end
       end
 
-      def set(key, value, application_id, username = CURRENT_USER, hostname = CURRENT_HOST)
-        CF.CFPreferencesSetValue(key.to_cf, value.to_cf, application_id.to_cf, arg_to_cf(username), arg_to_cf(hostname))
+      # Set the value for preference domain using `CFPreferencesSetValue`.
+      #
+      # @param [String] key Preference key
+      # @param [Integer, Float, String, TrueClass, FalseClass, Hash, Array] value Preference value
+      # @param [String] application_id Preference domain
+      # @param [String, Symbol] username Domain user (current, any or a specific user)
+      # @param [String, Symbol] hostname Hostname (current, all hosts or a specific host)
+      #
+      # @return [TrueClass, FalseClass] Returns true if preference was successfully written to storage, otherwise false.
+      #
+      def set(key, value, application_id, username = nil, hostname = nil)
+        if username.nil? or hostname.nil?
+          CF.CFPreferencesSetAppValue(
+            key.to_cf,
+            arg_to_cf(value),
+            application_id.to_cf
+          )
+        else
+          CF.CFPreferencesSetValue(
+            key.to_cf,
+            arg_to_cf(value),
+            application_id.to_cf,
+            arg_to_cf(username),
+            arg_to_cf(hostname)
+          )
+        end
         CF.CFPreferencesAppSynchronize(application_id.to_cf)
       end
 
-      def list_keys(application_id, username, hostname)
-        arr_ref = CF.CFPreferencesCopyKeyList(
-          application_id.to_cf,
-          arg_to_cf(username),
-          arg_to_cf(hostname)
-        )
-        CF::Array.new(arr_ref).to_ruby
+      # Calls the {#self.set} method and raise a `PreferenceSyncFailed` error if `false` is returned.
+      #
+      # @param [String] key Preference key to write
+      # @param [String] application_id Preference domain for the key
+      # @param [String, Symbol] username Domain user (current, any or a specific user)
+      # @param [String, Symbol] hostname Hostname (current, all hosts or a specific host)
+      #
+      # @raise [PreferenceSyncFailed] If {#self.set} call returned false.
+      #
+      # @return [VALUE] Returns nil if preference value is successfully written.
+      #
+      def set!(key, value, application_id, username = nil, hostname = nil)
+        raise(CF::Exceptions::PreferenceSyncFailed.new(key, application_id, hostname)) unless set(key, value, application_id, username, hostname)
```

Comment:
here also, we can just use `||`

Reply:
Doesn't have any functional difference. unless is more idiomatic.

Reply:
shortcircuit evaluation is different from using conditionals on multiple levels but even if you want to continue using condition I recommend 2 things -
1. Atleast make this and this -> https://github.com/chef/corefoundation/pull/21/files#diff-7a48df7397684c3a72b5e2e4bac4b585f806d0a0bb9d309eaecd54c1e06dd932L44 consistent
2. keep `action` before the `error` for readability, i.e. keep `set(key, value, application_id, username, hostname)` before `raise`

Reply:
1. `get` won't be using `||` because boolean is a valid type for domain properties. 
2. Trying to move the `set` call to left would make it a `if not` statement which according to the style guide should be replaced with `unless`.  

So let's go with what we have here.

Reply:
shortcircuit evaluation/control flow was good enough, https://github.com/rubocop/ruby-style-guide#if-as-a-modifier but since you have already changed, let's keep this